Improve Swagger Object creation

MathiusD opened the pull request:
Add creation of Swagger::Object by some object instance

(Related and insipred by #18)

Allow creation of Swagger::Object by something like that :
```crystal
struct User
  property id, nickname, username, email, bio

  def initialize(@id : String, @nickname : String, @username : String, @email : String, @bio : String? = nil)
  end
end

user = User.new(
  UUID.random.to_s, "icyleaf wang", "icyleaf", "icyleaf.cn@gmail.com", "Personal bio"
)

builder.add(Swagger::Object.create_from_instance(user))
```

P.S. : `src/swagger/http/handler.cr` was modified when the code formatting was executed via `crystal tool format src && crystal tool format spec`. I have therefore committed the formatting.

erdnaxeli committed:
Add ref support to Swagger::Property

MathiusD committed:
feat(swagger-object) : Improve object creation

Add creation of Swagger::Object by some object instance

(Related to #18)

icyleaf commented:
Thanks for the PR, I reviewed the code that the current implementation only contains the base type, so if there are complex types or nested variables such an implementation would be exceptionally complicated, my suggestion is to see if Crystal implements reflection natively.

MathiusD commented:
> Thanks for the PR, I reviewed the code that the current implementation only contains the base type, so if there are complex types or nested variables such an implementation would be exceptionally complicated, my suggestion is to see if Crystal implements reflection natively.

Indeed, I should have put this PR in Draft. I still have to do the following:

- [x] Manage classes and enums by reusing references (Cf #24)
  - [ ] Add spec for enums addition
- [ ] Add the missing primitive types
  - [ ] Add date
  - [ ] Add date-time
  - [ ] Add byte
  - [ ] Add binary
  - [ ] Add password ?

However I don't understand how the implementation would be harder without native reflection, indeed, for schemas the [open-api spec ](https://swagger.io/specification/#schema-object) the other schema and enums are pointed by reference. (In addition reflection made here via macros allows to get all the necessary information, in my opinion)

The only point which seems to me to pose problem intuitively is the management of the Union other than SomeClass? since we could not use it in the same way.

MathiusD committed:
feat(swagger-object) : Improve object creation by ref and add enum

Add management of enumeration
Add support of enumeration inside creation by instance of Swagger::Object
Add refs for other Swagger::Object for non primitive type in creation by instance

(Related to #18)

TODO : add spec for enumeration addition

MathiusD committed:
feat(swagger-object): Add management of array inside SwaggerDataType::create_from_class

MathiusD committed:
feat(swagger-object): Improve creation by instance and resolve ref

MathiusD committed:
feat(swagger-object): Return swagger_format as close as possible for integer case != Int32 or Int64

MathiusD committed:
fix(swagger-object): Patch spec for swagger object generation

MathiusD committed:
feat(swagger-object): name of instance is now compliant with instance directive when instance name is retrieved from class name

MathiusD committed:
feat(swagger-object): Add management of self references

MathiusD committed:
feat(swagger-array): Add management of array with primitives

MathiusD committed:
feat(swagger-object): Use JSON and YAML annotation of fields (Order of usage are : Json, Yaml and finally default comportment)

MathiusD committed:
feat(swagger-object): Add tests related to json and yaml field introspection

MathiusD committed:
feat(swagger-object): Test with array of primitives

MathiusD committed:
feat(swagger-object): Export primitive of array as primitive instead as ref
